### PR TITLE
Implemented all crud operations on author,publisher,category and book

### PR DESCRIPTION
**Pull request by Marchy2016**

### CODE REVIEW

File: src/main/java/com/roma/librarymanagment/controller/BookController.java [hunk 1]
```diff
@@ -84,16 +84,16 @@ private String saveBook(Model model, String isbn, Publisher publisher,Author aut
     @RequestMapping(path = "/deleteBook/{id}", method = RequestMethod.GET)
     private String deleteBook(@PathVariable Long id){
         bookService.deleteById(id);
-        return "menu";
+        return "redirect:/books";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/BookController.java]: Create a configuration properties class and add all this " "redirect:/books";

File: src/main/java/com/roma/librarymanagment/controller/BookController.java [hunk 2]
```diff
@@ -126,7 +126,6 @@ private String getBookById(Model model,@PathVariable Long id){
         Book book = bookService.findBookById(id);
         if(book != null){
             model.addAttribute("book", book);
-            System.out.println("Book" + book);
         }
         return "book";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/BookController.java:131]: Add in the configuration class

File: src/main/java/com/roma/librarymanagment/controller/CategoryController.java [hunk 3]
```diff
@@ -20,9 +21,9 @@ public CategoryController(CategoryService categoryService) {
 
     @RequestMapping(path = "/savecategory", method = RequestMethod.POST)
     public String addCategory(Model model,@ModelAttribute Category category){
-        final Category category1 = categoryService.add(category.getName());
-        model.addAttribute("categories",new ArrayList<Category>(){{add(category1);}});
-        return "menu";
+        final Category category1 = categoryService.addCategory(category.getName());
+        model.addAttribute("category",category1);
+        return "category";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/CategoryController.java:28]: add in configuration class

File: src/main/java/com/roma/librarymanagment/controller/CategoryController.java [hunk 4]
```diff
@@ -32,13 +33,38 @@ private String saveCategory(Model model){
...
+
+    @RequestMapping(path = "/updatecategory", method = RequestMethod.POST)
+    public String updateCategory(Model model, @ModelAttribute Category category){
+        Category category_ = categoryService.findCategoryById(category.getId());
+        if(category_ != null){
+            category_.setName(category.getName());
+            categoryService.updateCategory(category_);
+            model.addAttribute("category", category_);
+        }
+        return "category";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/CategoryController.java:68]: add in configuration class

File: src/main/java/com/roma/librarymanagment/controller/PublisherController.java [hunk 5]
```diff
@@ -23,26 +24,52 @@ public PublisherController(PublisherService publisherService) {
 
     @RequestMapping(path = "/publishers", method = RequestMethod.GET)
     private String listPublishers(Model model){
-        model.addAttribute("publishers", publisherService.findAll());
+        model.addAttribute("publishers", publisherService.listPublishers());
         return "publishers";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/PublisherController.java:27]: add in configuration class

File: src/main/java/com/roma/librarymanagment/controller/PublisherController.java [hunk 6]
```diff
@@ -23,26 +24,52 @@ public PublisherController(PublisherService publisherService) {
...
 
     @RequestMapping(path = "/savepublishers", method = RequestMethod.POST)
-    private String addPublisher(Model model, @ModelAttribute Publisher publisher){
-        Publisher publisher1 = publisherService.add(publisher.getName(),publisher.getAddress(),publisher.getEmail());
-        model.addAttribute("publisher", new ArrayList<Publisher>(){{add(publisher1);}});
-        return "menu";
+    private String savePublisher(Model model, @ModelAttribute Publisher publisher){
+        Publisher publisher1 = publisherService.addPublisher(publisher.getName(),publisher.getAddress(),publisher.getEmail());
+        model.addAttribute("publisher", publisher1);
+        return "publisher";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/PublisherController.java:42]: add in configuration class

File: src/main/java/com/roma/librarymanagment/controller/PublisherController.java [hunk 7]
```diff
@@ -23,26 +24,52 @@ public PublisherController(PublisherService publisherService) {
...
+        Publisher publisher1 = publisherService.updatePublisher(publisher.getId(),publisher.getName(),publisher.getAddress(),publisher.getEmail());
+        model.addAttribute("publisher", publisher1);
+        return "publisher";
+
+    }
+
+    @RequestMapping(path = "/deletepublisher/{id}", method = RequestMethod.GET)
+    private String deletePublisher(Model model, @PathVariable Long id){
+          publisherService.deletePublisher(id);
+        return "redirect:/publishers";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/PublisherController.java]: add in configuration class

File: src/main/java/com/roma/librarymanagment/controller/PublisherController.java [hunk 8]
```diff
@@ -23,26 +24,52 @@ public PublisherController(PublisherService publisherService) {
...
+          if(publisher != null){
+              model.addAttribute("publisher",publisher);
+          }
+        return "updatePublisher";
+    }
+    @RequestMapping(path = "/updatepublisher", method = RequestMethod.POST)
+    private String updatePublisher(Model model, @ModelAttribute Publisher publisher){
+        Publisher publisher1 = publisherService.updatePublisher(publisher.getId(),publisher.getName(),publisher.getAddress(),publisher.getEmail());
+        model.addAttribute("publisher", publisher1);
+        return "publisher";
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/controller/PublisherController.java:59]: add in configuration class

File: src/main/java/com/roma/librarymanagment/repositories/AuthorRepository.java [hunk 9]
```diff
@@ -2,8 +2,11 @@
 
 import com.roma.librarymanagment.model.Author;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 public interface AuthorRepository extends JpaRepository<Author, Long> {
+    @Query("update Author set firstName = firstName, lastName = lastName, email = email where id = id")
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/repositories/AuthorRepository.java:9]: change the the code here and  use proper springdata query language

File: src/main/java/com/roma/librarymanagment/repositories/AuthorRepository.java [hunk 10]
```diff
@@ -2,8 +2,11 @@
...
+    Author updateAuthor(Long id, String firstName, String lastName,String email);;
```
Comment (sonwabo) [inline src/main/java/com/roma/librarymanagment/repositories/AuthorRepository.java]: add the @Param annotation to the arguments

**Review by sonwabo — COMMENTED** [review]
Please re-look at your changes

File: src/main/java/com/roma/librarymanagment/controller/BookController.java (hunk 1 above)
Comment (Marchy2016) [reply src/main/java/com/roma/librarymanagment/controller/BookController.java]: Can you please clarify or point me to sites where I can clarity on what you requesting.

Comment (sonwabo) [reply src/main/java/com/roma/librarymanagment/controller/BookController.java]: https://www.baeldung.com/configuration-properties-in-spring-boot

**Pull request merged**